File: frontend/src/shared/capsule/fieldTable.ts
/**
 * The 145-bit capsule field table — CLAUDE.md, "Capsule" section. This is the
 * single source of truth for field order and width; encode() and decode() both
 * derive offsets from it so they cannot drift apart (capsule-auditor's first
 * check: "Recompute offsets from widths rather than trusting the constants").
 *
 * Total width must be exactly 145 (75 free inside the 220-bit NavIC segment).
 * `capsule.test.ts` asserts this on import, not just in CI.
 */

export interface FieldSpec {
  name: string;
  width: number;
}

// Order and widths transcribed verbatim from CLAUDE.md's field table.
export const FIELD_TABLE: readonly FieldSpec[] = [
  { name: "msg_type", width: 4 },
  { name: "schema_ver", width: 3 },
  { name: "issue_slot", width: 10 },
  { name: "valid_hours", width: 4 },
  { name: "zone_id", width: 20 },
  { name: "lat_offset", width: 8 },
  { name: "lon_offset", width: 8 },
  { name: "risk_class", width: 2 },
  { name: "hazard_flags", width: 8 },
  { name: "wave_hs", width: 5 },
  { name: "wind_kt", width: 6 },
  { name: "wind_dir", width: 4 },
  { name: "curr_kt", width: 5 },
  { name: "curr_dir", width: 4 },
  { name: "sst", width: 5 },
  { name: "chl_class", width: 2 },
  { name: "pfz_bearing", width: 4 },
  { name: "pfz_dist_nm", width: 6 },
  { name: "pfz_conf", width: 2 },
  { name: "bnd_dist_nm", width: 5 },
  { name: "bnd_type", width: 2 },
  { name: "bnd_eta_min", width: 8 },
  { name: "reason_code", width: 8 },
  { name: "evidence_hash", width: 12 },
] as const;

export const CAPSULE_TOTAL_BITS = 145;
export const NAVIC_SEGMENT_BITS = 220;
export const CAPSULE_BYTES = 19; // ceil(145/8) = 19, matches the QR/BLE/SMS wire size

export interface FieldOffset extends FieldSpec {
  offset: number;
}

/** Cumulative offsets, recomputed every time — never hand-maintained. */
export function fieldOffsets(): FieldOffset[] {
  let offset = 0;
  return FIELD_TABLE.map((f) => {
    const withOffset = { ...f, offset };
    offset += f.width;
    return withOffset;
  });
}

export function assertTableIntegrity(): void {
  const offsets = fieldOffsets();
  const total = offsets.reduce((sum, f) => sum + f.width, 0);
  if (total !== CAPSULE_TOTAL_BITS) {
    throw new Error(
      `capsule field table is ${total} bits, expected exactly ${CAPSULE_TOTAL_BITS}`,
    );
  }
  const names = new Set<string>();
  for (const f of offsets) {
    if (names.has(f.name)) throw new Error(`duplicate capsule field: ${f.name}`);
    names.add(f.name);
  }
  // No field may overlap another — true by construction from cumulative
  // offsets, but checked explicitly so a future hand-edit cannot break it.
  for (let i = 1; i < offsets.length; i++) {
    const prevEnd = offsets[i - 1].offset + offsets[i - 1].width;
    if (offsets[i].offset !== prevEnd) {
      throw new Error(`capsule field ${offsets[i].name} does not start where the prior field ends`);
    }
  }
}

// Fails fast on import if the table is ever hand-edited into an inconsistent state.
assertTableIntegrity();

File: backend/dhruva/codec/field_table.py
"""The 145-bit capsule field table.

Must stay byte-for-byte identical in meaning to
frontend/src/shared/capsule/fieldTable.ts — `tests/test_codec_corpus.py` and
the frontend's `codec.test.ts` both run
`eval/capsule_corpus.json` and must agree on every capsule in it. If this
table and the TypeScript one ever diverge, that shared corpus test is what
catches it, not code review.
"""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True, slots=True)
class FieldSpec:
    name: str
    width: int


FIELD_TABLE: tuple[FieldSpec, ...] = (
    FieldSpec("msg_type", 4),
    FieldSpec("schema_ver", 3),
    FieldSpec("issue_slot", 10),
    FieldSpec("valid_hours", 4),
    FieldSpec("zone_id", 20),
    FieldSpec("lat_offset", 8),
    FieldSpec("lon_offset", 8),
    FieldSpec("risk_class", 2),
    FieldSpec("hazard_flags", 8),
    FieldSpec("wave_hs", 5),
    FieldSpec("wind_kt", 6),
    FieldSpec("wind_dir", 4),
    FieldSpec("curr_kt", 5),
    FieldSpec("curr_dir", 4),
    FieldSpec("sst", 5),
    FieldSpec("chl_class", 2),
    FieldSpec("pfz_bearing", 4),
    FieldSpec("pfz_dist_nm", 6),
    FieldSpec("pfz_conf", 2),
    FieldSpec("bnd_dist_nm", 5),
    FieldSpec("bnd_type", 2),
    FieldSpec("bnd_eta_min", 8),
    FieldSpec("reason_code", 8),
    FieldSpec("evidence_hash", 12),
)

CAPSULE_TOTAL_BITS = 145
NAVIC_SEGMENT_BITS = 220
CAPSULE_BYTES = 19  # ceil(145/8)


@dataclass(frozen=True, slots=True)
class FieldOffset(FieldSpec):
    offset: int


def field_offsets() -> list[FieldOffset]:
    offset = 0
    out: list[FieldOffset] = []
    for f in FIELD_TABLE:
        out.append(FieldOffset(f.name, f.width, offset))
        offset += f.width
    return out


def _assert_table_integrity() -> None:
    offsets = field_offsets()
    total = sum(f.width for f in offsets)
    if total != CAPSULE_TOTAL_BITS:
        raise AssertionError(f"capsule field table is {total} bits, expected exactly {CAPSULE_TOTAL_BITS}")
    seen: set[str] = set()
    for f in offsets:
        if f.name in seen:
            raise AssertionError(f"duplicate capsule field: {f.name}")
        seen.add(f.name)
    prev_end = 0
    for f in offsets:
        if f.offset != prev_end:
            raise AssertionError(f"capsule field {f.name} does not start where the prior field ends")
        prev_end = f.offset + f.width


_assert_table_integrity()

WIDTH_BY_NAME: dict[str, int] = {f.name: f.width for f in field_offsets()}
